include/aspect/velocity_boundary_conditions/ascii_data.h

```diff
@@ -0,0 +1,519 @@
+            */
+           const unsigned int components;
+
+           /**
+            * A reference to the geometry model. Is needed to convert the
+            * position into spherical coordinates if necessary.
+            */
+           const GeometryModel::Interface<dim> &geometry_model;
+
+           /**
+            * Interpolation functions to access the data.
+            */
+           std::vector<Functions::InterpolatedUniformGridData<grid_dim> *> data;
+           std::vector<Functions::InterpolatedUniformGridData<grid_dim> *> old_data;
+
+           /**
+            * Model size
+            */
+           std_cxx11::array<std::pair<double,double>,grid_dim> grid_extent;
+
+           /**
+            * Number of points in the data grid.
+            */
+           TableIndices<grid_dim> table_points;
+
+           /**
+            * Dimensions of the boundary plane
+            */
+           unsigned int boundary_dimensions[grid_dim];
+
+           /**
+            * Scale the data boundary condition by a scalar factor. Can be
+            * used to transform the unit of the data.
+            */
+           const double scale_factor;
+
+
+           std_cxx11::array<std::pair<double,double>,dim>
+           get_model_extent (const GeometryModel::Interface<dim> &geometry_model) const
+           {
+             std_cxx11::array<std::pair<double,double>,dim> data_extent;
+
+             if (dynamic_cast<const GeometryModel::Box<dim>*> (&geometry_model) != 0)
+               {
+                 const GeometryModel::Box<dim> *geometry_model_box = dynamic_cast<const GeometryModel::Box<dim> *>(&geometry_model);
+                 const Point<dim> box_origin = geometry_model_box->get_origin();
+                 const Point<dim> box_upper_bound = box_origin + geometry_model_box->get_extents();
+
+                 for (unsigned int i=0;i<dim;i++)
+                   data_extent[i] = std::make_pair<double,double> (box_origin[i],box_upper_bound[i]);
+               }
+             else if (dynamic_cast<const GeometryModel::SphericalShell<dim>*> (&geometry_model) != 0)
+               {
+                 const GeometryModel::SphericalShell<dim> *geometry_model_spherical_shell =
+                     dynamic_cast<const GeometryModel::SphericalShell<dim> *>(&geometry_model);
+
+                 data_extent[0] = std::make_pair(0,geometry_model_spherical_shell->maximal_depth());
+                 data_extent[1] = std::make_pair(0,geometry_model_spherical_shell->opening_angle() * numbers::PI / 180);
+                 if (dim==3)
+                   data_extent[2] = std::make_pair<double,double> (0,geometry_model_spherical_shell->opening_angle() * numbers::PI / 180);
+               }
+             else Assert (false, ExcNotImplemented());
+
+             return data_extent;
+           }
+
+           std_cxx11::array<unsigned int,grid_dim>
+           get_boundary_dimensions (const types::boundary_id boundary_id) const
```

Comment:
This function determines which of the dimensions of the position is used to find the data point in the data grid. E.g. the left boundary of a box model extents in the y, and z direction (position[1] and position[2]), therefore get_boundary_dimensions would return [1,2] for dim==3 or [1] for dim==2. We are lucky that these indices are identical for the box and the spherical shell (if we use spherical coordinates for the spherical shell), therefore we do not need to distinguish between them. For the initial condition this function is trivial, because the position in the data grid is the same as the actual position (the function returns [0,1,2]  or [0,1]), but for the boundary conditions it matters.
But you are right, this function deserves a detailed comment.